core/coreapi/interface/interface.go

```diff
@@ -73,6 +105,16 @@ type CoreAPI interface {
 	// ResolveNode resolves the path (if not resolved already) using Unixfs
 	// resolver, gets and returns the resolved Node
 	ResolveNode(context.Context, Path) (Node, error)
+
+	// ParsePath parses string path to a Path
+	ParsePath(context.Context, string, ...options.ParsePathOption) (Path, error)
+
+	// WithResolve is an option for ParsePath which when set to true tells
+	// ParsePath to also resolve the path
+	WithResolve(bool) options.ParsePathOption
```

Comment:
I'm wondering if I should make it default to true. cc @lgierth

Reply:
This should definitely not be a method on the CoreAPI interface (just a function, or maybe just a type `type WithResolve bool`). However, I notice we're already doing this all over the place. Any reason for it?

Anyways, do we ever actually *need* `ResolvePath` to resolve? Doesn't everything that takes a path do resolution internally just in case?

Reply:
It's based on a concept called 'functional options' from [this article](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis).

TL;DR - this pattern let's us expand the options without breaking anything, it might be overkill for some methods, but I prefer consistency more. But yeah, a separate method for resolving string path makes sense too

> Anyways, do we ever actually need ResolvePath to resolve? Doesn't everything that takes a path do resolution internally just in case?

The problem is that paths are (currently) immutable and if user wanted to re-use the resolved path (ipns can be slow), we'd need to return the resolved path, which would result in calls to some methods getting a bit messy.

Reply:
Having an explicit resolve will also make `go-ipfs-api` implementation simpler.

Reply:
> It's based on a concept called 'functional options' from this article.

I like the pattern, I'm just objecting to attaching these functions to the interface itself instead of leaving them as free functions. It actually goes against the spirit of the pattern as it makes it *harder* to grow the API: adding new options will break existing implementations of the CoreAPI interface.

Reply:
> The problem is that paths are (currently) immutable and if user wanted to re-use the resolved path (ipns can be slow), we'd need to return the resolved path, which would result in calls to some methods getting a bit messy.

Fair enough. However, in this case (and many others) we *know* that we're only going to use the path once (as far as I can tell).

Reply:
> I like the pattern, I'm just objecting to attaching these functions to the interface itself instead of leaving them as free functions.

I feel exactly the same.